src/asl.ts

```diff
@@ -2918,6 +2934,285 @@ export class ASL {
     );
   }
 
+  public evalDecl(
+    decl: VariableDecl | ParameterDecl,
+    initialValue?: ASLGraph.Output
+  ): ASLGraph.SubState | ASLGraph.NodeState | undefined {
+    const getState = () => {
+      if (initialValue) {
+        return this.evalAssignment(decl.name, initialValue);
+      } else if (decl.initializer === undefined) {
+        return undefined;
+      }
+
+      return this.evalExprToSubState(decl.initializer, (exprOutput) => {
+        return this.evalAssignment(decl.name, exprOutput);
+      });
+    };
+    const state = getState();
+    return state ? { node: decl, ...state } : undefined;
+  }
+
+  /**
+   * Generic handler for any type of assignment with either an identifier or a binding pattern (deconstruction).
+   *
+   * Nested BindPatterns are handled recursively.
+   *
+   * const x = 1;
+   * =>
+   * const x = 1;
+   *
+   * const { x } = y;
+   * =>
+   * const x = y.x;
+   *
+   * const { x } = { x: 1 };
+   * =>
+   * const x = 1;
+   *
+   * const { x = 1 } = y;
+   * =>
+   * const x = y.x === undefined ? 1 : y.x;
+   *
+   * const { x = 1 } = {};
+   * =>
+   * const x = 1;
+   *
+   * const { x: { z } } = y;
+   * =>
+   * const z = y.x.z;
+   *
+   * const { x, ...rest } = y;
+   * =>
+   * INVALID - rest is unsupported in objects because ASL doesn't support object manipulation (ex: delete) or field enumeration (ex: keySet).
+   *
+   * const [x] = arr;
+   * =>
+   * const x = arr[0];
+   *
+   * const [,x] = arr;
+   * =>
+   * const x = arr[1];
+   *
+   * const [x,...rest] = arr;
+   * =>
+   * const x = arr[0];
+   * const rest = arr.slice(1);
+   */
+  private evalAssignment(
+    pattern: Identifier | BindingPattern,
+    value: ASLGraph.Output
+  ): ASLGraph.SubState | ASLGraph.NodeState | undefined {
+    // assign an identifier the current value
+    if (isIdentifier(pattern)) {
+      return ASLGraph.passWithInput(
+        {
+          Type: "Pass",
+          ResultPath: `$.${pattern.name}`,
+        },
+        value
+      );
+    } else {
+      const rest = pattern.bindings.find((binding) => binding?.rest);
+      if (rest && isObjectBinding(pattern)) {
+        // TODO: validator
+        throw new SynthError(
+          ErrorCodes.StepFunctions_does_not_support_destructuring_object_with_rest
+        );
+      }
+
+      // run each of the assignments as a sequence of states, they should not rely on each other.
```

Comment:
Can default values point to other bound names?

Reply:
uhh, I guess they can... I didn't know that was possible.

```ts
const { a, b = a } = { a: "1" }
b // '1'
```

Well I guess I have more work to do :-P

Though in theory this should work because the resulting names will be public and in order

```ts
const { c = d, d } = { d: "1" }

// Uncaught ReferenceError: d is not defined
```

Will add tests

Reply:
Worked!